File: apps/clinical-engine/src/ehas2_clinical_engine/rule4/selection/d3_d5_discriminator_fingerprint_v1.py
from __future__ import annotations

import hashlib
from typing import Any

from ..canonical_json import canonical_stable_dumps

from .d3_d5_evidence_provenance import evidence_provenance_fingerprint_payload

RULE4_D3D5_DISCRIMINATOR_FINGERPRINT_V1 = "rule4-d3d5-discriminator-fingerprint-v1"


def _gate_payload(gate: dict) -> dict:
    return {
        "gate_id": gate["gate_id"],
        "outcome": gate["outcome"],
        "evidence_item_ids": sorted(gate.get("evidence_item_ids") or []),
        "reason_codes": sorted(gate.get("reason_codes") or []),
        "limitation_codes": sorted(gate.get("limitation_codes") or []),
    }


def rule4_d3d5_discriminator_fingerprint_v1_payload(envelope: dict) -> str:
    gates = sorted(envelope.get("q7bf_gate_results") or [], key=lambda g: g["gate_id"])
    return canonical_stable_dumps(
        {
            "fingerprint_version": RULE4_D3D5_DISCRIMINATOR_FINGERPRINT_V1,
            "formula_slot_id": envelope["formula_slot_id"],
            "formula_target_id": envelope["formula_target_id"],
            "upstream_phase7_eligibility_fingerprint": envelope[
                "upstream_phase7_eligibility_fingerprint"
            ],
            "q7bf_gate_results": [_gate_payload(g) for g in gates],
            "d08_document_status": envelope["d08_document_status"],
            "d08_document_confidence": envelope["d08_document_confidence"],
            "d08_item_status": envelope["d08_item_status"],
            "d08_item_confidence": envelope["d08_item_confidence"],
            "sensitivity_assessment_status": envelope["sensitivity_assessment_status"],
            "close_d05_discriminator_status": envelope["close_d05_discriminator_status"],
            "evidence_item_ids": sorted(envelope.get("evidence_item_ids") or []),
            "evidence_provenance": evidence_provenance_fingerprint_payload(
                list(envelope.get("evidence_provenance") or [])
            ),
            "source_reference_ids": sorted(envelope.get("source_reference_ids") or []),
            "binding_status": envelope["binding_status"],
        }
    )


def rule4_d3d5_discriminator_fingerprint_v1_hash(envelope: dict) -> str:
    payload = rule4_d3d5_discriminator_fingerprint_v1_payload(envelope)
    return hashlib.sha256(payload.encode("utf-8")).hexdigest().upper()


def with_computed_discriminator_fingerprint(envelope: dict) -> dict:
    body = {k: v for k, v in envelope.items() if k != "discriminator_fingerprint"}
    return {**body, "discriminator_fingerprint": rule4_d3d5_discriminator_fingerprint_v1_hash(body)}


def discriminator_fingerprint_fields(envelope: dict) -> dict[str, Any]:
    return {
        "formula_slot_id": envelope["formula_slot_id"],
        "formula_target_id": envelope["formula_target_id"],
        "upstream_phase7_eligibility_fingerprint": envelope[
            "upstream_phase7_eligibility_fingerprint"
        ],
        "q7bf_gate_results": envelope.get("q7bf_gate_results") or [],
        "d08_document_status": envelope["d08_document_status"],
        "d08_document_confidence": envelope["d08_document_confidence"],
        "d08_item_status": envelope["d08_item_status"],
        "d08_item_confidence": envelope["d08_item_confidence"],
        "sensitivity_assessment_status": envelope["sensitivity_assessment_status"],
        "close_d05_discriminator_status": envelope["close_d05_discriminator_status"],
        "evidence_item_ids": envelope.get("evidence_item_ids") or [],
        "evidence_provenance": envelope.get("evidence_provenance") or [],
        "source_reference_ids": envelope.get("source_reference_ids") or [],
        "binding_status": envelope["binding_status"],
    }

File: packages/clinical-contracts/src/rule4/selection/d3D5DiscriminatorFingerprintV1.ts
import { createHash } from 'node:crypto';

import { canonicalStableDumps } from '../canonicalJson.js';
import type { Rule4GateResult } from '../eligibility/types.js';
import { evidenceProvenanceFingerprintPayload } from './d3D5EvidenceProvenance.js';
import type {
  Rule4CloseD05DiscriminatorStatus,
  Rule4D08ConfidenceStatus,
  Rule4D08UsabilityStatus,
  Rule4D3D5DiscriminatorEnvelope,
  Rule4D3D5SensitivityAssessmentStatus,
} from './types.js';

export const RULE4_D3D5_DISCRIMINATOR_FINGERPRINT_V1 =
  'rule4-d3d5-discriminator-fingerprint-v1' as const;

function gatePayload(gate: Rule4GateResult): Record<string, unknown> {
  return {
    gate_id: gate.gateId,
    outcome: gate.outcome,
    evidence_item_ids: [...gate.evidenceItemIds].sort(),
    reason_codes: [...gate.reasonCodes].sort(),
    limitation_codes: [...gate.limitationCodes].sort(),
  };
}

export function rule4D3D5DiscriminatorFingerprintV1Payload(
  envelope: Omit<Rule4D3D5DiscriminatorEnvelope, 'discriminatorFingerprint'>,
): string {
  const gates = [...envelope.q7bfGateResults]
    .sort((a, b) => a.gateId.localeCompare(b.gateId))
    .map(gatePayload);
  return canonicalStableDumps({
    fingerprint_version: RULE4_D3D5_DISCRIMINATOR_FINGERPRINT_V1,
    formula_slot_id: envelope.formulaSlotId,
    formula_target_id: envelope.formulaTargetId,
    upstream_phase7_eligibility_fingerprint: envelope.upstreamPhase7EligibilityFingerprint,
    q7bf_gate_results: gates,
    d08_document_status: envelope.d08DocumentStatus,
    d08_document_confidence: envelope.d08DocumentConfidence,
    d08_item_status: envelope.d08ItemStatus,
    d08_item_confidence: envelope.d08ItemConfidence,
    sensitivity_assessment_status: envelope.sensitivityAssessmentStatus,
    close_d05_discriminator_status: envelope.closeD05DiscriminatorStatus,
    evidence_item_ids: [...envelope.evidenceItemIds].sort(),
    evidence_provenance: evidenceProvenanceFingerprintPayload(envelope.evidenceProvenance ?? []),
    source_reference_ids: [...envelope.sourceReferenceIds].sort(),
    binding_status: envelope.bindingStatus,
  });
}

export function rule4D3D5DiscriminatorFingerprintV1Hash(
  envelope: Omit<Rule4D3D5DiscriminatorEnvelope, 'discriminatorFingerprint'>,
): string {
  const payload = rule4D3D5DiscriminatorFingerprintV1Payload(envelope);
  return createHash('sha256').update(payload, 'utf8').digest('hex').toUpperCase();
}

export function withComputedDiscriminatorFingerprint(
  envelope: Omit<Rule4D3D5DiscriminatorEnvelope, 'discriminatorFingerprint'>,
): Rule4D3D5DiscriminatorEnvelope {
  return {
    ...envelope,
    discriminatorFingerprint: rule4D3D5DiscriminatorFingerprintV1Hash(envelope),
  };
}

export type Rule4D3D5DiscriminatorFingerprintFields = {
  formulaSlotId: string;
  formulaTargetId: string;
  upstreamPhase7EligibilityFingerprint: string;
  q7bfGateResults: readonly Rule4GateResult[];
  d08DocumentStatus: Rule4D08UsabilityStatus;
  d08DocumentConfidence: Rule4D08ConfidenceStatus;
  d08ItemStatus: Rule4D08UsabilityStatus;
  d08ItemConfidence: Rule4D08ConfidenceStatus;
  sensitivityAssessmentStatus: Rule4D3D5SensitivityAssessmentStatus;
  closeD05DiscriminatorStatus: Rule4CloseD05DiscriminatorStatus;
  evidenceItemIds: readonly string[];
  evidenceProvenance: readonly import('./types.js').Rule4D3D5EvidenceProvenanceRecord[];
  sourceReferenceIds: readonly string[];
  bindingStatus: Rule4D3D5DiscriminatorEnvelope['bindingStatus'];
};
